src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractTranslator.java

```diff
@@ -335,6 +339,29 @@ public String translate(Update u) {
         return join(temp, " ");
     }
 
+    public String translate(final With with) {
+
+        final List<ImmutablePair<Name, Expression>> withs = with.getWiths();
+
+        withs.forEach(aWith -> {
+            injector.injectMembers(aWith.getLeft());
+            injector.injectMembers(aWith.getRight());
+        });
+
+        final String withStatements = withs.stream()
+                .map(aWith -> aWith.getLeft().translate() + " AS (" + aWith.getRight().translate() + ")")
+                .collect(Collectors.joining(", "));
+
+        final Expression then = with.getThen();
+
+        if (then != null) {
```

Comment:
Does it makes sense to have a WITH without the "then"?  
What is an example usage of that?

Reply:
It does not make sense. However, enforcing the declaration of `then` statement, before all the "withs" are declared, seems strange.
`with("a", <query>, <then-query>);`
What do you think?

Reply:
Please keep it the way it is.